components/engine_rocks_helper/src/sst_recovery.rs

```diff
@@ -0,0 +1,245 @@
+                    || (!keys::validate_data_key(&f.largestkey)
+                        && f.largestkey.as_slice() != keys::DATA_MAX_KEY)
+                {
+                    set_panic_mark_and_panic(
+                        path,
+                        &format!(
+                            "key range mismatch, small key:{:?}, large key:{:?}",
+                            &f.smallestkey, &f.largestkey
+                        ),
+                    );
+                }
+
+                // lock the store_meta and check if the range exist.
+                if !self.check_overlap_damaged_regions(f) {
+                    set_panic_mark_and_panic(path, "Can't find overlap regions");
+                }
+
+                return;
+            }
+        }
+    }
+
+    // `sst_path` has been processed and is still in a damaged state.
+    fn exist_scheduling_regions(&self, sst_path: &str) -> bool {
+        self.damaged_files.iter().any(|f| f.name == sst_path)
+    }
+
+    // Periodically check for recorded damaged files。
+    //
+    // Acquire meta lock.
+    fn check_damaged_files(&mut self) {
+        for index in 0..self.damaged_files.len() {
+            if self.damaged_files[index].start_time.elapsed() > self.max_damage_duration {
+                let f = self.damaged_files.remove(index);
+                let fname = f.name.clone();
+                // file may be put into `damaged_files` again here, but when this happens,
+                // panic will be triggered, so it doesn’t matter.
+                if self.check_overlap_damaged_regions(f) {
+                    set_panic_mark_and_panic(
+                        &fname,
+                        &format!("recovery job exceeded {:?}", self.max_damage_duration),
+                    );
+                }
+            }
+        }
+    }
+
+    // Check whether the StoreMeta contains the region range, if it contains,
+    // recorded fault region ids to report to PD and add file info into `damaged_files`.
+    //
+    // Acquire meta lock.
+    fn check_overlap_damaged_regions(&mut self, mut file: FileInfo) -> bool {
+        let mut meta = match self.store_meta.lock() {
+            Ok(meta) => meta,
+            Err(e) => {
+                set_panic_mark_and_panic(&file.name, &format!("{}", e));
+                return false;
+            }
+        };
+
+        let mut overlap = false;
+
+        // Find overlapping region ids.
+        // Condition:
+        // end_key > file.smallestkey
+        // start_key <= file.largestkey
+        let mut ids = vec![];
+        for (_, id) in meta
+            .region_ranges
+            .range((Excluded(file.smallestkey.clone()), Unbounded::<Vec<u8>>))
+        {
+            let region = &meta.regions[id];
+            if keys::enc_start_key(region) <= file.largestkey {
+                ids.push(*id);
+                file.overlap_region_ids.push(*id);
+                overlap = true;
+            }
```

Comment:
else break?

Reply:
This requires implicit ordering guarantees and there doesn't seem to be any performance impact without breaking early here(handling damaged regions is not a normal operation)

Reply:
What's the ordering guarantee? The `range()` is a guarantee